gnosis-imap/gnosis-imap-spec.ini

```diff
@@ -148,6 +148,158 @@ proxy_storage:
+   andBool M1 ==IMap M2 except (
+        SetItem(#hashedLocation({COMPILER}, {OWNERS}, OWNER))
+        SetItem(#hashedLocation({COMPILER}, {OWNERS}, PREV_OWNER))
+        SetItem(#hashedLocation({COMPILER}, {THRESHOLD}, .IntList))
+        SetItem(#hashedLocation({COMPILER}, {OWNERCOUNT}, .IntList)) 
+        .Set)
+
+[version2-removeOwner-success]
+k: (#execute => #halt) ~> _
+output: _ => _
+statusCode: _ => EVMC_SUCCESS
+callStack: _
+this: #PROXY_ID
+msg_sender: #PROXY_ID
+callData: #abiCallData("removeOwner", (#address(PREV_OWNER), 
+                                       #address(OWNER), 
+                                       #uint256(NEW_THRESHOLD)))
+callValue: 0
+wordStack: .WordStack => _
+localMem: .IMap => _
+pc: 0 => _
+gas: #gas(STARTGAS, 0, 0) => _
+memoryUsed: 0 => _
+log: _:List ( .List => ListItem(#abiEventLog(#PROXY_ID, "RemovedOwner", #address(OWNER))) )
+refund: _ => _
+coinbase: _ => _
+proxy_storage:
+    M1:IMap => M2:IMap
++requires:
+    ; Address public constant SENTINEL_OWNERS = address(0x1);
+    andBool SENTINEL ==Int 1
+    ; Range
+    andBool #rangeAddress(  PREV_OWNER)
+    andBool #rangeAddress(  OWNER)
+    andBool #rangeAddress(  INIT_OWNER)
+    andBool #rangeAddress(  INIT_PREV_OWNER)
+    andBool #rangeUInt(256, NEW_THRESHOLD)
+    andBool #rangeUInt(256, THRESHOLD)  
+    andBool #rangeUInt(256, OWNERCOUNT)
+    andBool #rangeUInt(256, OWNERCOUNT -Int 1)
+    ; Storage
+    andBool select(M1, #hashedLocation({COMPILER}, {OWNERS}, OWNER))        ==Int INIT_OWNER
+    andBool select(M1, #hashedLocation({COMPILER}, {OWNERS}, PREV_OWNER))   ==Int INIT_PREV_OWNER
+    andBool select(M1, #hashedLocation({COMPILER}, {THRESHOLD}, .IntList))  ==Int THRESHOLD
+    andBool select(M1, #hashedLocation({COMPILER}, {OWNERCOUNT}, .IntList)) ==Int OWNERCOUNT
+    ; Requirements from code
+    ; Only allow to remove an owner, if threshold can still be reached.
+    andBool OWNERCOUNT -Int 1 >=Int NEW_THRESHOLD
+    ; Validate owner address and check that it corresponds to owner index
+    andBool OWNER =/=Int 0
+    andBool OWNER =/=Int SENTINEL
+    andBool INIT_PREV_OWNER ==Int OWNER
+    ; Threshold is unchanged.
+    andBool THRESHOLD ==Int NEW_THRESHOLD 
+ ;   ; path condition
+    andBool OWNER =/=Int PREV_OWNER     // implied from 'owners[prevOwner] == owner' and 'owner != sentinel'
+ensures:
+   ensures true
+   andBool select(M2, #hashedLocation({COMPILER}, {OWNERS}, OWNER))         ==Int 0
+   andBool select(M2, #hashedLocation({COMPILER}, {OWNERS}, PREV_OWNER))    ==Int INIT_OWNER
+   andBool select(M2, #hashedLocation({COMPILER}, {OWNERCOUNT}, .IntList))  ==Int OWNERCOUNT -Int 1 
+   andBool M1 ==IMap M2 except (
+        SetItem(#hashedLocation({COMPILER}, {OWNERS}, OWNER))
+        SetItem(#hashedLocation({COMPILER}, {OWNERS}, PREV_OWNER))
+        SetItem(#hashedLocation({COMPILER}, {OWNERCOUNT}, .IntList)) 
+        .Set)
+
```

Comment:
I cannot see how this version works.  It may be due to z3 magic, but I'm not sure if this approach will work for other function specs.

Reply:
As we are pressured by time, I think it is better just to drop this alternative approach and use the one you already reviewed. In any case, I saved this approach and I'll come back to it when we are done with the Gnosis project.